### PR TITLE
MNIST Doc

### PR DESCRIPTION
**Pull request by codeboy5**
Hi, @kartikdutt18 I have made the changes, have a look.
Created a new PR due to some issues.

### CODE REVIEW

File: Kaggle/README.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+<i>A visualization of the sample image at index 7777.<i/>
+</p>
+
+<br/>
+
+### Processing The Dataset
+
+##### 1. Loading The Dataset From The File To A Useable Format
+
+The Load Class is used to load a matrix from the file.The type of file is guessed automatically from the extension. For a list of supported file types please refer to [Load Class](https://github.com/mlpack/mlpack/blob/master/src/mlpack/core/data/load.hpp). Please note that since mlpack requires matrices to be in column major format and usually data is stored in a row major format, the matrix by default is transposed. In case the data is originally in column major format, consider setting the parameter to False. Refer to documentation of Load Class for further information.
```
Comment (favre49) [inline Kaggle/README.md]:
```suggestion
The Load Class is used to load a matrix from the file. The type of file is guessed automatically from the extension. For a list of supported file types please refer to [Load Class](https://github.com/mlpack/mlpack/blob/master/src/mlpack/core/data/load.hpp). Please note that since mlpack requires matrices to be in column major format and usually data is stored in a row major format, the matrix is transposed by default. In case the data is originally in column major format, consider setting `transpose` to false. Refer to documentation of Load Class for further information.
```

Can we link to the official documentation instead of the file directly?

File: Kaggle/README.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+After loading the dataset is stored in the _tempDataset_ Matrix.
+
+```cpp
+mat tempDataset;
+data::Load("Kaggle/data/train.csv", tempDataset, true);
+```
+
+##### 2. Splitting The Dataset
+
+The Split Class is used to split the Dataset into Train and Validation Datasets. The train dataset is used in the training procedure, the weights are updated and adjusted to fit this dataset. The validation dataset is used to constantly evaludate the model, which is used to adjust the model hyperparamters like the learning rate. The data is split randomly and RATIO defines the amount of data in Validation Set.
```
Comment (favre49) [inline Kaggle/README.md]:
```suggestion
The Split Class is used to split the dataset into training and validation datasets. The training dataset is used in the training procedure. The weights are updated and adjusted to fit this dataset. The validation dataset is used to constantly evaluate the model and adjust the model hyperparameters like the learning rate. The data is split randomly and RATIO defines the amount of data in the validation dataset.
```

File: Kaggle/README.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+After splitting the train dataset is stored in the _train_ matrix and validation in the _valid_ matrix.
```
Comment (favre49) [inline Kaggle/README.md]:
```suggestion
After splitting the training dataset is stored in the _train_ matrix and validation dataset in the _valid_ matrix.
```

File: Kaggle/README.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+```cpp
+  mat train, valid;
+  data::Split(dataset, train, valid, RATIO);
+```
+
+### Results Obtained on Different Models
+
+##### 1. DigitRecognizer
+
+For More Details related to the Model, check out the **DigitRecognizer** Directory.
```
Comment (favre49) [inline Kaggle/README.md]:
```suggestion
For more details related to the model, check out the **DigitRecognizer** Directory.
```
Can we link to the directory? Could you also explain what these models actually are?

The same suggestion for everything below

File: Kaggle/README.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+After running the model for 50 cycles, the model obtained the best validation set accuracy of **97.2143%**.
```
Comment (favre49) [inline Kaggle/README.md]:
```suggestion
After running the model for 50 epochs, the model obtained a validation set accuracy of **97.2143%**.
```

Same suggestions below.

File: Kaggle/README.md (hunk 4 above)
Comment (codeboy5) [reply Kaggle/README.md]: So the idea we discussed (@kartikdutt18) was to explain the dataset and processing along with the results in this README and explain the models in their respective folders, e.g. DigitRecognizerCNN would contain a description of the CNN model.

File: Kaggle/README.md (hunk 1 above)
Comment (codeboy5) [reply Kaggle/README.md]: Sure makes sense.

File: Kaggle/README.md (hunk 4 above)
Comment (favre49) [reply Kaggle/README.md]: Just a quick overview, so people know what they're looking at. Needn't be in depth at all.

Comment (codeboy5) [reply Kaggle/README.md]: Ok, I will do that.

Comment (codeboy5) [reply Kaggle/README.md]: Is something like this model is a NN with 2 hidden layers and NLL as a loss function or should i add more ?

Comment (favre49) [reply Kaggle/README.md]: No, I was thinking something more like what Ryan wrote [here](https://github.com/mlpack/examples/blob/92385adcca7cf117875a03cd370dfd0d7e2699a1/README.md)

Comment (kartikdutt18) [reply Kaggle/README.md]: Hi @codeboy5, @favre49, So this is what I think, this repo is also being restructured to remove folders and each model will have its own folder so I believe this is what should be done:
1. Have a notebook like structure that makes more sense for repo like this.

So this would follow this step:

1. A basic overview of dataset (probably an image), shape of image and labels.

2. Steps to implement the model. For example:

>
Step 3: Create the model
We can create a CNN model by stacking Convolutional layer as shown below.
```
   model->Add(layer1);
    ..
   model->Add(layerN);
```

No need to add all layers just the show 1 or 2 cn layers.

> Step 4: Training
One or two lines about using ensmallen callbacks with code.

and so.

Let me know if this makes sense.

Comment (kartikdutt18) [reply Kaggle/README.md]: I'll open documentation for LSTM today you can check that out and maybe that makes sense.

Comment (favre49) [reply Kaggle/README.md]: Ah, I forgot about this. However, wouldn't it be better in some cases to group some examples together? Like for Digit Recognizers, or for LSTM (basically, like the current file structure). I'm ambivalent on this, but it does fix a problem for when we have, say, 40 or 50 examples. When repos get big, I personally like it when they are well organized

Comment (kartikdutt18) [reply Kaggle/README.md]: That's what I had thought as well. I thought we could have folders like:
```
--mnist_object_classification
  --cnn
  --nn
  --batchnorm
--timeseries_univariate
  --lstm
--timeseries_multivariate
  --lstm
```
This also makes documentation cleaner and comprehensive.
What do you think?
 I think this would also need to be reflected in #69 then.

Comment (favre49) [reply Kaggle/README.md]: Yeah I was thinking the same too. Though IMO we can have timeseries programs under an LSTM examples folder instead.

Comment (codeboy5) [reply Kaggle/README.md]: Waiting for the documentation of LSTM, will edit this to look similar

Comment (kartikdutt18) [reply Kaggle/README.md]: Hey Saksham, A slight delay will push the docs first thing tomorrow, Finally figured out the bug in another PR so I am wrapping that up. Sorry to keep you waiting.

Comment (codeboy5) [reply Kaggle/README.md]: No problems

File: Kaggle/README.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
+# MNIST
+
+### DESCRIPTION OF THE DATASET
+
+MNIST dataset contains images of handwritten digits. It is one of the most common datasets used for image classifcation. It has 60,000 28x28 grayscale images under the training set and 10,000 28x28 grayscale images under the test set. Each pixel has a value between 0 and 255.
```
Comment (birm) [inline Kaggle/README.md:5]:
```suggestion
The MNIST dataset contains images of handwritten digits. It is one of the most common datasets used for image classifcation. It has 60,000 28x28 grayscale images under the training set and 10,000 28x28 grayscale images under the test set. Each pixel has a value between 0 and 255.
```

File: Kaggle/README.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
+# MNIST
+
+### DESCRIPTION OF THE DATASET
```
Comment (birm) [inline Kaggle/README.md:3]: Please be consistent with capitalization of headings.

File: Kaggle/README.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+MNIST dataset contains images of handwritten digits. It is one of the most common datasets used for image classifcation. It has 60,000 28x28 grayscale images under the training set and 10,000 28x28 grayscale images under the test set. Each pixel has a value between 0 and 255.
+
+<p  align="center">
+
+<img  alt="Sample Images From The MNIST Database"  src="https://upload.wikimedia.org/wikipedia/commons/2/27/MnistExamples.png">
+
+</p>
+<p  align="center">
+<i>Sample Images From The MNIST Database.<i/>
```
Comment (birm) [inline Kaggle/README.md:13]:
```suggestion
<i>Sample Images From The MNIST Database.</i>
```

File: Kaggle/README.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+</p>
+
+<br/>
+
+<p  align="center">
+
+<img  alt="A visualization of the sample image at index 7777"  src="https://miro.medium.com/max/490/1*nlfLUgHUEj5vW7WVJpxY-g.png">
```
Comment (birm) [inline Kaggle/README.md:20]: Is this a stable url?

File: Kaggle/README.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+For more details related to the model, check out the **DigitRecognizer** Directory.
+
+After running the model for 50 epochs, the model obtained a validation set accuracy of **97.2143%**.
+
+##### 2. DigitRecognizerBatchNorm
+
+For More Details related to the Model, check out the **DigitRecognizerBatchNorm** Directory.
+
+After running the model for 50 cycles, the model obatained the best validation set accuracy of **91.5952%**.
```
Comment (birm) [inline Kaggle/README.md:64]: I'm probably not up to date on terminology, but I'm not sure what you mean by cycles here? Iterations? Epochs? Some other thing?

File: Kaggle/README.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+</p>
+
+<br/>
+
+<p  align="center">
+
+<img  alt="A visualization of the sample image at index 7777"  src="https://miro.medium.com/max/490/1*nlfLUgHUEj5vW7WVJpxY-g.png">
+</p>
+<p  align="center">
+<i>A visualization of the sample image at index 7777.<i/>
```
Comment (birm) [inline Kaggle/README.md:23]:
```suggestion
<i>A visualization of the sample image at index 7777.</i>
```

File: Kaggle/README.md (hunk 10 above)
Comment (codeboy5) [reply Kaggle/README.md:64]: Epochs

File: Kaggle/README.md [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+##### 1. Loading The Dataset From The File To A Useable Format
+
+The Load Class is used to load a matrix from the file. The type of file is guessed automatically from the extension. For a list of supported file types please refer to [Load Class](https://github.com/mlpack/mlpack/blob/master/src/mlpack/core/data/load.hpp). Please note that since mlpack requires matrices to be in column major format and usually data is stored in a row major format, the matrix is transposed by default. In case the data is originally in column major format, consider setting `transpose` to false. Refer to documentation of Load Class for further information.
+
+After loading the dataset is stored in the _tempDataset_ Matrix.
+
+```cpp
+mat tempDataset;
+data::Load("Kaggle/data/train.csv", tempDataset, true);
```
Comment (shrit) [inline Kaggle/README.md:38]:
```suggestion
```cpp
mat tempDataset;
data::Load("../data/train.csv", tempDataset, true);
```

File: Kaggle/README.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+```cpp
+  mat train, valid;
+  data::Split(dataset, train, valid, RATIO);
+```
+
+### Results Obtained on Different Models
+
+##### 1. DigitRecognizer
+
+For more details related to the model, check out the **DigitRecognizer** Directory.
```
Comment (shrit) [inline Kaggle/README.md:56]:
```suggestion
For more details related to the model, check out the **mnist** Directory.
```

File: Kaggle/README.md [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+After running the model for 50 epochs, the model obtained a validation set accuracy of **97.2143%**.
+
+##### 2. DigitRecognizerBatchNorm
+
+For More Details related to the Model, check out the **DigitRecognizerBatchNorm** Directory.
```
Comment (shrit) [inline Kaggle/README.md:62]:
```suggestion
For More Details related to the Model, check out the **mnist_batch_norm** Directory.
```

File: Kaggle/README.md [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+### Results Obtained on Different Models
+
+##### 1. DigitRecognizer
+
+For more details related to the model, check out the **DigitRecognizer** Directory.
+
+After running the model for 50 epochs, the model obtained a validation set accuracy of **97.2143%**.
+
+##### 2. DigitRecognizerBatchNorm
```
Comment (shrit) [inline Kaggle/README.md:60]:
```suggestion
##### 2. Mnist Batch Norm
```

File: Kaggle/README.md [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+After splitting the training dataset is stored in the _train_ matrix and validation dataset in the _valid_ matrix.
+
+```cpp
+  mat train, valid;
+  data::Split(dataset, train, valid, RATIO);
+```
+
+### Results Obtained on Different Models
+
+##### 1. DigitRecognizer
```
Comment (shrit) [inline Kaggle/README.md:54]:
```suggestion
##### 1. Mnist
```

File: Kaggle/README.md [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+
+##### 2. DigitRecognizerBatchNorm
+
+For More Details related to the Model, check out the **DigitRecognizerBatchNorm** Directory.
+
+After running the model for 50 cycles, the model obatained the best validation set accuracy of **91.5952%**.
+
+##### 3. DigitRecognizerCNN
+
+For More Details related to the Model, check out the **DigitRecognizerCNN** Directory.
```
Comment (shrit) [inline Kaggle/README.md:68]:
```suggestion
##### 3. Mnist CNN

For More Details related to the Model, check out the **mnist_cnn** Directory.
```

File: Kaggle/README.md (hunk 9 above)
Comment (birm) [reply Kaggle/README.md:20]: Well, it's been a few months and still looks ok.

**Comment by codeboy5** [comment]
> Hey @codeboy5, thanks for taking the time to write this up. In the time since you opened the PR, the directory structure of the repository has changed somewhat. So if we do merge this, we should be sure that the directory we use is more appropriate since `Kaggle/` doesn't exist anymore. Maybe `mnist_simple`? It's hard to say though, since there are several directories with MNIST models.
> 
> Personally I wonder if this would be more effective as inline comments in each of the `mnist_*.cpp` example files. Eventually, when we have notebooks for each of these examples, the inline images will make a lot more sense. (For now we can just include URLs in the comments.) My worry is just that if this is a separate file, people might not see it---but if it's integrated into the code directly, they have to see it. 😄
> 
> (I'm requesting changes for the directory structure issue---from my side that's the main thing I'd like to see change before merging this, but hopefully my other points are useful for discussion too. 👍)

I will make the changes and update it

**Comment by shrit** [comment]
@codeboy5 are you interested in making changes?

**Comment by shrit** [comment]
keep open
